### ISSUE TITLE
Advanced Auditing 1.9 umbrella bug

### ISSUE THREAD

**Issue by crassirostris** [kubernetes/kubernetes#54551]
This is a continuation of the work on the [Advanced Auditing feature](https://github.com/kubernetes/features/issues/22), that was tracked for 1.8 release in https://github.com/kubernetes/kubernetes/issues/48561

As discussed earlier, in 1.9 release API stays in Beta for stabilization. Here's the list of tasks for this K8s release:

## API-related changes

- [x] Add a policy-wide `OmitStages` field (more context in https://github.com/kubernetes/kubernetes/pull/49280#issuecomment-319711384)
  - @CaoShuFeng is working on this in https://github.com/kubernetes/kubernetes/pull/54634
- [x] Fix the timestamps in the API (https://github.com/kubernetes/kubernetes/issues/52160)
  - @CaoShuFeng is working on this in https://github.com/kubernetes/kubernetes/pull/52981

## Pipeline bugfixes

- [x] Make webhook parameters configurable
  - @crassirostris is working on this in https://github.com/kubernetes/kubernetes/pull/56638 and https://github.com/kubernetes/kubernetes/pull/56890
- [x] ~Introduce buffering for the log backend to fix file writing bottleneck (https://github.com/kubernetes/kubernetes/issues/53006)~
  - Deferred until the next release
- [x] ~Restore audit logging in the scalability tests (https://github.com/kubernetes/kubernetes/issues/53020)~
  - Deferred until the next release
- [x] Shutdown http handlers before shutting down audit backend (https://github.com/kubernetes/kubernetes/issues/50781)
  - @hzxuzhonghu is working on this in https://github.com/kubernetes/kubernetes/pull/54849
- [x] Policy without kind/apiVersion is still valid (https://github.com/kubernetes/kubernetes/issues/54254)
  - @ericchiang is working on this in https://github.com/kubernetes/kubernetes/pull/54267
- [x] Rate-limit batching webhook backend (https://github.com/kubernetes/kubernetes/issues/52907)
  - @crassirostris is working on this in https://github.com/kubernetes/kubernetes/pull/53417, with the help of @hzxuzhonghu
- [x] Network errors are not retried (https://github.com/kubernetes/kubernetes/issues/52909)
  - @crassirostris is working on this in https://github.com/kubernetes/kubernetes/pull/53947

## Policy changes

- [x] ~GCE audit policy should be made re-usable by other setups (https://github.com/kubernetes/kubernetes/issues/53321)~
  - Deferred until the next release
- [x] ~Audit policy should be tested (e.g. that it includes all core resources, more context in https://github.com/kubernetes/kubernetes/issues/52265#issuecomment-329243972)~
  - Deferred until the next release

## Misc

- [x] kubectl sends empty patch requests, which results in addon manager spamming audit logs (https://github.com/kubernetes/kubernetes/issues/54010)
  - @hzxuzhonghu is working on this in https://github.com/kubernetes/kubernetes/pull/54046

## To discuss

- [x] ~Auditing federation setups (https://github.com/kubernetes/kubernetes/issues/50076)~
  - Deferred until the next release
- [x] ~Auditing multi-apiserver setups~
  - Deferred until the next release

/cc @sttts @soltysh @tallclair @ericchiang @CaoShuFeng @hzxuzhonghu

**Comment by crassirostris** [comment]
Feel free to add things I forgot or correct existing items

**Comment by CaoShuFeng** [comment]
Do we still need this?
https://github.com/kubernetes/kubernetes/pull/49280#issuecomment-319711384

If so, I will do it.

**Comment by crassirostris** [comment]
@CaoShuFeng Thanks! That would help, added to the list of tasks

**Comment by dims** [comment]
/assign @CaoShuFeng
/assign @crassirostris 

@CaoShuFeng @crassirostris hope you are ok as assignees for this issue. please unassign/reassign as appropriate

**Comment by jberkus** [comment]
Note from the release team: This issue is marked Approved-for-Milestone 1.9.  However, many of the associated PRs are not approved.  Code Slush is Nov. 20th; do you think the PRs will be complete and approved by then, or should this be moved out of the milestone?

**Comment by crassirostris** [comment]
@jberkus Sorry, I was preempted by another effort and haven't updated the issue. It's actually is much better shape

One API-related change is close to approval, another is simple and will also make it until the cutoff

Enhancements and bug fixes AFAIU don't need to be approved on November 20th, do they?

**Comment by crassirostris** [comment]
OK, clarifications: what's left is a big bug that is important for this milestone: synchronously logging to disk is a huge bottleneck that doesn't allow to enable file-based audit logging in large clusters

I think this is not a part of the feature work and can be done after the feature freeze

**Comment by tallclair** [comment]
Should the the audit API move into the `k8s.io/api` repo?

@sttts and @soltysh say no (https://github.com/kubernetes/kubernetes/pull/45315#issuecomment-300700982), but I wonder if you still feel that way? Counter arguments include:
- audit log file defaults to a json representation of the API, so it is useful for parsing that log
- Not a client, but the target of the webhook may not be an actual apiserver, and can use the API
- https://github.com/kubernetes/kubernetes/issues/53455 proposes a streaming audit endpoint, which could be directly consumed by a client.

**Comment by crassirostris** [comment]
This actually sounds reasonable, I agree with Tim.

To add to his point, I was thinking recently about an opensource component that would listen for the audit webhook on the master machine and would push audit logs to an external system that doesn't understand K8s API (e.g. Elasticsearch).  Having audit API available in the client would make it easier to implement such component.

@sttts @soltysh Do you have objections now?

**Comment by crassirostris** [comment]
The last remaining pieces to be done are critical bugs, so bumping the priority to critical

**Comment by crassirostris** [comment]
Still blocked on https://github.com/kubernetes/kubernetes/issues/53020
ETA: 1/12/2017
Risks: Complicated fix required

**Comment by liggitt** [comment]
If this is blocked on an issue with eta of 1/12, it shouldn't be in the 1.9 milestone

**Comment by crassirostris** [comment]
@liggitt To clarify: the issue is about making it possible to load test the feature, which is ready since the previous release. There's no feature work left & it doesn't make sense to disable the feature on 2017-12-01, since it's there since 1.8

**Comment by liggitt** [comment]
Sure, but the continued work is targeting 1.10, not 1.9 at this point, so the milestone should change.

**Comment by crassirostris** [comment]
@liggitt Dec 01, not Jan 12, right?

**Comment by enisoc** [comment]
@crassirostris Please file an [exception request](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) for any remaining work on this. It sounds like the performance problem has been there since 1.8, so it's unlikely that we would block the rest of 1.9 on that problem. I may be wrong on that, but either way there is enough doubt around this that the exception process is the right place to make the case and decide.

**Comment by crassirostris** [comment]
@enisoc Sure, done

**Comment by crassirostris** [comment]
As discussed in the [exception requests thread](https://groups.google.com/forum/#!topic/kubernetes-milestone-burndown/zdzq-7zWMZM), we're not proceeding with buffering for this milestone, instead, I'll send a PR to make webhook parameters configurable.

**Comment by crassirostris** [comment]
All things in the core K8s are addressed. There's one change left in the GCE cluster configuration, I'll address it in a separate PR

**Comment by jberkus** [comment]
ETA for the GCE request?

**Comment by crassirostris** [comment]
@jberkus Today

**Comment by crassirostris** [comment]
https://github.com/kubernetes/kubernetes/pull/56890 closes this issue
